File: petclinic/src/main/java/com/petclinic/pet/PetService.java
package com.petclinic.pet;

import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class PetService {
    
    private final PetRepository petRepository;
    
    public PetService(PetRepository petRepository) {
        this.petRepository = petRepository;
    }
    
    public Optional<Pet> findById(Long id) {
        return petRepository.findById(id);
    }
    
    public Optional<Pet> findByName(String name) {
        return petRepository.findByName(name);
    }
    
    public List<Pet> findAll() {
        return petRepository.findAll();
    }
    
    public Pet save(Pet pet) {
        return petRepository.save(pet);
    }
}

File: brownfield-backend/src/pet/petService.ts
import { Pet } from "./pet";
import { PetRepository } from "./petRepository";

export class PetService {
  private petRepository: PetRepository;

  constructor(petRepository?: PetRepository) {
    this.petRepository = petRepository ?? new PetRepository();
  }

  findAll(): Promise<Pet[]> {
    return this.petRepository.findAll();
  }

  findById(id: number): Promise<Pet | null> {
    return this.petRepository.findById(id);
  }

  findByName(name: string): Promise<Pet | null> {
    return this.petRepository.findByName(name);
  }

  async save(pet: Pet): Promise<Pet> {
    await this.validatePetUniquenessPerOwner(pet);
    return this.petRepository.save(pet);
  }

  async delete(id: number): Promise<void> {
    return this.petRepository.delete(id);
  }

  private async validatePetUniquenessPerOwner(pet: Pet): Promise<void> {
    const existing = await this.petRepository.findByNameAndOwnerName(
      pet.name,
      pet.ownerName,
    );
    if (existing !== null && existing.id !== pet.id) {
      throw new Error(
        `Owner ${pet.ownerName} already has a pet named ${pet.name}`,
      );
    }
  }
}
